src/program_iterator.jl

```diff
@@ -67,56 +67,112 @@ end
-            Expr(:kw, :(max_depth::Int), typemax(Int)), 
-            Expr(:kw, :(max_size::Int), typemax(Int)), 
-            Expr(:kw, :(max_time::Int), typemax(Int)), 
-            Expr(:kw, :(max_enumerations::Int), typemax(Int)),
-            Expr(:kw, :(solver::Union{Solver, Nothing}), nothing)
-        ]
+        kwargs_fields = map(esc, filter(is_kwdef, extrafields))
+        notkwargs     = map(esc, filter(!is_kwdef, extrafields))
+
+        # create field names
+        field_names = map(extract_name_from_argument, extrafields)
+
+        # throw an error if user used one of the reserved arg names 
+        RESERVERD_ARG_NAMES = [:solver,:start_symbol,:initial_node,:grammar,:max_depth,:max_size]
+        for field_name ∈ field_names
+            if field_name ∈ RESERVERD_ARG_NAMES
+                throw(ArgumentError(
+                    "When using the @programiterator macro you are not allowed to use any of the $RESERVERD_ARG_NAMES field names.
+                     This is because there would be conflicting names in the function signature. 
+                     However, '$field_name' was found as an argument name. 
+                     Please change the name of the field argument to not collide with the reserved argument names above.
+                    "))
+            end
+        end
 
+        field_names = map(esc, field_names)
+        escaped_name = esc(name) # this is the name of the struct
+
+        # keyword arguments come after the normal arguments (notkwargs)
+        all_constructors = Base.remove_linenums!(
+            :(
+              begin 
+                # solver with grammar and start symbol
+                function $(escaped_name)(grammar::AbstractGrammar, start_symbol::Symbol, $(notkwargs...) ; 
+                                        max_size = typemax(Int), max_depth = typemax(Int), $(kwargs_fields...) )
+                    return $(escaped_name)(GenericSolver(grammar, start_symbol, max_size = max_size, max_depth = max_depth), $(field_names...))
+                end
+
+                # solver with grammar and initial rulenode to start with
+                function $(escaped_name)(grammar::AbstractGrammar, initial_node::RuleNode, $(notkwargs...) ;
+                                        max_size = typemax(Int), max_depth = typemax(Int), $(kwargs_fields...) )
+                    return $(escaped_name)(GenericSolver(grammar, initial_node, max_size = max_size, max_depth = max_depth), $(field_names...))
+                end
+              end
+            )
+        )
+        # this constructor should ONLY be used when there are kwarg fields 
+        # otherwise this will overwrite the default julia struct constructor
+        solver_constructor = Base.remove_linenums!(:(
+          # solver main constructor
+          function $(escaped_name)(solver::Solver, $(notkwargs...) ; $(kwargs_fields...) )
+              return $(escaped_name)(solver, $(field_names...))
+          end
+        ))
```

Comment:
```suggestion
        solver_constructor = Base.remove_linenums!(:(
          # solver main constructor
          function $(escaped_name)(solver::Solver, $(notkwargs...) ; max_size = nothing, max_depth = nothing, $(kwargs_fields...) )
              if !isnothing(max_size) solver.max_size = max_size end
              if !isnothing(max_depth) solver.max_depth = max_depth end
              return $(escaped_name)(solver, $(field_names...))
          end
        ))
```

Reply:
Ok I think it makes sense. I will commit this